### PR TITLE
Use DeepDiff library for a/b record diffing

### PR DESCRIPTION
**Pull request by ghukill**
### Purpose and background context

Originally, the library `jsondiff` was used for getting a diff between the A and B version of a JSON record.  This was selected based on prior use and experience.  For the most part it was a nice fit, but lacked the ability to ignore changes in array order as a difference to report.  This resulted in high numbers of records showing a diff, when the only change was a differently ordered array.

The solution was to utilize another JSON diffing library, [`DeepDiff`](https://zepworks.com/deepdiff/8.0.1/diff.html#module-deepdiff.diff).

How this addresses that need:
* `DeepDiff` library has built-in support to `ignore_order` when generating diffs, allowing us to programatically ignore order in arrays as a diff to report
* `DeepDiff` is as fast, or faster, than `jsondiff` making it a performant option
* Our primary, current use of diffs is to highlight records where a TIMDEX field changed, which is conveniently a "root" field on the JSON object.  `DeepDiff` provides the attribute [`.affected_root_keys`](https://zepworks.com/deepdiff/current/diff.html#deepdiff.diff.DeepDiff.affected_root_keys) on the diff object that returns an explicit list of modified "root" fields, which is just what we need, further simplifying some logic in ABDiff to parse this information.  [Here is where that is applied in our code](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/47/files#diff-f6d81437fc3dd591a024246e6ea6c0b3093e76e4c16e3cb4a018736c22bf0924R111).  Note in [this commit](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/47/commits/edc5f80dda71028adfc35bd2e658751054102634) how additional diff parsing logic is no longer needed.

Overall, `DeepDiff` is a more configurable library, with built-in options that align closely with our use case.

### How can a reviewer manually see the effects of these changes?

See the following tests to understand how array order is an optional difference to report (where we default to _not_ reporting it):
- [`test_calc_record_diff_array_by_default_order_not_a_diff`](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/47/files#diff-8ba6561e4eaa024fa152a30999873753c6ab7c68c351e28ca427ef7e3902e5b8R42-R49)
- [`test_calc_record_diff_array_set_flag_order_is_a_diff`](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/47/files#diff-8ba6561e4eaa024fa152a30999873753c6ab7c68c351e28ca427ef7e3902e5b8R52-R57)

### Includes new or updated dependencies?
YES

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-373

### Developer
- [X] All new ENV is documented in README
- [X] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: abdiff/core/calc_ab_diffs.py [hunk 1]
```diff
@@ -61,23 +63,52 @@ def get_diffed_batches_iter(
...
+        df["has_diff"] = diff_results.apply(lambda x: x[2])
 
         yield pa.RecordBatch.from_pandas(df)  # type: ignore[attr-defined]
 
 
-def calc_record_diff(record_a: bytes | None, record_b: bytes | None) -> str | None:
-    """Calculate symmetric diff from two JSON strings."""
+def calc_record_diff(
+    record_a: str | bytes | dict | None,
+    record_b: str | bytes | dict | None,
```
Comment (ehanson8) [inline abdiff/core/calc_ab_diffs.py:82]: Why the type hint change? It feel odd to account 3 different potential types for these `record_*` params, is there any way to normalize around one?

File: abdiff/core/calc_ab_metrics.py [hunk 2]
```diff
@@ -66,32 +64,31 @@ def create_record_diff_matrix_dataset(
...
+                "has_diff": (1 if row["has_diff"] == "true" else 0),
             }
-            diff_data = row["ab_diff"]
-            record_metrics.update(generate_field_diff_bools_for_record(diff_data))
+
+            # for each modified field (root key in diff), set column and value = 1 (True)
+            if row["modified_timdex_fields"] is not None:
+                for field in row["modified_timdex_fields"]:
+                    record_metrics[field] = 1
+
```
Comment (ehanson8) [inline abdiff/core/calc_ab_metrics.py:85]: This reads a lot cleaner, `deepdiff` seems like a great change

File: tests/test_calc_ab_diffs.py [hunk 3]
```diff
@@ -17,22 +17,66 @@
...
+
+
+def test_calc_record_diff_array_by_default_order_not_a_diff():
+    """Arrays with the same values, but differently ordered, not considered a diff."""
+    a = {"colors": ["green", "red"]}
+    b = {"colors": ["red", "green"]}
+    ab_diff, modified_timdex_fields, has_diff = calc_record_diff(a, b)
+    assert ab_diff == json.dumps({})  # no diff
+    assert not modified_timdex_fields
+    assert not has_diff
```
Comment (ehanson8) [inline tests/test_calc_ab_diffs.py:49]: Loving that we can have this test already, I thought it was going to be a far more involved fix

File: tests/test_calc_ab_diffs.py [hunk 4]
```diff
@@ -17,22 +17,66 @@
...
+
+def test_calc_record_diff_array_set_flag_order_is_a_diff():
+    """Arrays with the same values, but differently ordered, not considered a diff."""
+    a = {"colors": ["green", "red"]}
+    b = {"colors": ["red", "green"]}
+    _, _, has_diff = calc_record_diff(a, b, ignore_order=False)
+    assert has_diff
+
+
+def test_calc_record_diff_array_repetition_is_reported_when_diff():
```
Comment (ehanson8) [inline tests/test_calc_ab_diffs.py:60]: Great edge cases to highlight in the unit tests

File: tests/test_calc_ab_metrics.py [hunk 5]
```diff
@@ -12,32 +12,10 @@
...
-    diff_data = {
-        "$insert": {"fruits": "strawberry"},
-        "$delete": {"vegetables": "onion"},
-    }
-    assert generate_field_diff_bools_for_record(diff_data) == {
-        "fruits": 1,
-        "vegetables": 1,
-    }
-
-
```
Comment (ehanson8) [inline tests/test_calc_ab_metrics.py:40]: "Cutting is delivering" is my new favorite motto

**Review by ehanson8 — COMMENTED** [review]
One question, but this is a great update!

File: abdiff/core/calc_ab_diffs.py (hunk 1 above)
Comment (ghukill) [reply abdiff/core/calc_ab_diffs.py:82]: Originally, the type signature was `bytes | None`.  We either got bytes when a record had the version (and it usually did) or handle situations where it was missing.  

Supporting `dict` as well made [testing easier](https://github.com/MITLibraries/transmogrifier-ab-diff/blob/edc5f80dda71028adfc35bd2e658751054102634/tests/test_calc_ab_diffs.py#L18-L20) where we didn't have to constantly convert a dictionary to bytes, just to have this convert it back.

While we are expecting `bytes` from the input datasets, got to thinking there is no reason why it _can't_ also support `str` as well.  For a future developer, I think it should represent what it can accept without extra work or modification, and this includes `str` as well.

It's a good question, but in this instance, I don't think we benefit by limiting the allowed types to this function.  I think it's fairly common in conversion or serialization type libraries, where they can handle _lots_ of types of inputs generally.

File: abdiff/core/calc_ab_metrics.py (hunk 2 above)
Comment (ghukill) [reply abdiff/core/calc_ab_metrics.py:85]: 100% agree!  The fact it picks out the modified "root" keys, works perfectly with our TIMDEX records.

File: tests/test_calc_ab_metrics.py (hunk 5 above)
Comment (ghukill) [reply tests/test_calc_ab_metrics.py:40]: Hahahahah, agreed!  No need for veggies.

File: abdiff/core/calc_ab_metrics.py [hunk 6]
```diff
@@ -107,32 +104,6 @@ def create_record_diff_matrix_dataset(
...
-            for subfield in diff_data[key]:
-                fields_with_diffs[subfield] = 1
-
-        # identified modified fields at root of diff
-        else:
-            fields_with_diffs[key] = 1
-
-    return fields_with_diffs
-
-
```
Comment (jonavellecuerdo) [inline abdiff/core/calc_ab_metrics.py:135]: Woohoo! ✨

**Review by jonavellecuerdo — APPROVED** [review]
Great find! Awesome job. 🎉
